samples/99_zil_interpreter/app/parser.rb

```diff
@@ -0,0 +1,417 @@
+
+    i = 0
+    while i < num
+      str += @buffer[@buf_index + i]
+      i += 1
+    end
+
+    str
+  end
+
+  def is_fix?(char)
+    return false if char.nil?
+
+    if FIX_CHARS.include?(char)
+      true
+    else
+      false
+    end
+  end
+
+  # ,ATOM
+  def read_alias_gval
+    skip_char(',')
+    expr = read_expression
+
+    Syntax::Form.new(:GVAL, expr)
+  end
+
+  # .ATOM
+  def read_alias_lval
+    skip_char('.')
+    expr = read_expression
+
+    Syntax::Form.new(:LVAL, expr)
+  end
+
+  def read_atom
+    atom = ""
+
+    while @buf_index < @buffer.length - 1
+      buf_char = peek
+      buffer_advance
+
+      if buf_char == '\\'
+        atom += read_char
+      elsif ATOM_CHARS.include?(buf_char)
+        atom += buf_char
+      else
+        buffer_previous
+        break
+      end
+    end
+
+    raise "Invalid syntax at #{@line_num}:#{@line_pos}!" if atom.length == 0
+
+    atom.to_sym
+  end
+
+  # [start]...[end]
+  def read_bracket_expression(start_bracket, end_bracket, output_class)
+    skip_whitespace
+    skip_char(start_bracket)
+    skip_whitespace
+
+    # read contents
+    expr = []
+    next_char = peek
+    while next_char != end_bracket
+      expr << read_expression
+      skip_whitespace
+      next_char = peek
+    end
+
+    skip_char(end_bracket)
+
+    output_class.new(*expr)
+  end
+
+  def read_char
+    char = peek
+    buffer_advance
+    char
+  end
+
+  # !\.
+  def read_character
+    skip_char('!')
+    skip_char('\\')
+    read_char
+  end
+
+  # ;...
+  def read_comment
+    # read comment char, then read the expresson that follows
+    skip_char(';') # read in ;
+    skip_whitespace
+    comment = read_expression # read commented out stuff and ignore
+
+    Syntax::Comment.new(comment)
+  end
+
+  # #DECL(...)
+  def read_decl
+    skip_char('#') # read #
+    skip_atom(:DECL) # read DECL
+
+    list = read_list
+    raise "DECL inner list is empty!" if list.elements.length < 1
+
+    Syntax::Decl.new(*list.elements) # return inner list as convenience
+  end
+
+  def read_expression
+    debug_flag = false
+    @expr_depth += 1
+    raise "Expression recursion overflow" if @expr_depth > 50
+
+    expr_char = peek
+    if expr_char == '<' #  Starts with '<' then read_form
+      log indent(@expr_depth) + "+ FORM " if debug_flag
+      expr = read_form
+    elsif expr_char == '(' #  Starts with '(' then read_list
+      log indent(@expr_depth) + "+ LIST " if debug_flag
+      expr = read_list
+    elsif expr_char == '"' #  Starts with '"' then read_string
+      log indent(@expr_depth) + "+ STRING " if debug_flag
+      expr = read_string
+    elsif expr_char == '!' && peek_next == '\\' #  Starts with '!\' then read_character
+      log indent(@expr_depth) + "+ CHARACTER " if debug_flag
+      expr = read_character
+    elsif expr_char == '!' && peek_next == '\\' #  Starts with '!' then read_segment
+      log indent(@expr_depth) + "+ SEGMENT " if debug_flag
+      expr = read_segment

```

Comment:
the two statements have the same condition if you look closely.... the latter one should probably be
```rb
elsif expr_char == '!' && peek_next != '\\'
```
so `!= '\\'` instead of `== '\\'`

But how about summarizing this into:
```suggestion
    elsif expr_char == '!'
      if peek_next == '\\' #  Starts with '!\' then read_character
        log indent(@expr_depth) + "+ CHARACTER " if debug_flag
        expr = read_character
      else #  Starts with '!' then read_segment
        log indent(@expr_depth) + "+ SEGMENT " if debug_flag
        expr = read_segment
      end
```
?

Reply:
Fixed.